net.go

```diff
@@ -118,6 +120,233 @@ func (e *EurekaConnection) GetApps() (map[string]*Application, error) {
 	return apps, nil
 }
 
+func instanceCount(apps []*Application) int {
+	count := 0
+	for _, app := range apps {
+		count += len(app.Instances)
+	}
+	return count
+}
+
+type instanceQueryOptions struct {
+	// secure indicates whether to query a secure or an insecure VIP address.
+	secure bool
+	// predicate guides filtering, indicating whether to retain an instance when it returns true or
+	// drop it when it returns false.
+	predicate func(*Instance) bool
+	// intn behaves like the rand.Rand.Intn function, aiding in randomizing the order of the result
+	// sequence when non-nil.
+	intn func(int) int
+}
+
+// InstanceQueryOption is a customization supplied to instance query functions like
+// GetInstancesByVIPAddress to tailor the set of instances returned.
+type InstanceQueryOption func(*instanceQueryOptions) error
+
+func retainIfStatusIs(status StatusType, o *instanceQueryOptions) {
+	if prev := o.predicate; prev != nil {
+		o.predicate = func(instance *Instance) bool {
+			return prev(instance) || instance.Status == status
+		}
+	} else {
+		o.predicate = func(instance *Instance) bool {
+			return instance.Status == status
+		}
+	}
+}
+
+// WithStatus restricts the set of instances returned to only those with the given status.
+//
+// Supplying multiple options produced by this function applies their logical disjunction.
+func WithStatus(status StatusType) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		if len(status) == 0 {
+			return errors.New("invalid instance status")
+		}
+		retainIfStatusIs(status, o)
+		return nil
+	}
+}
+
+func makeUncheckedStatusMatchingOption(status StatusType) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		retainIfStatusIs(status, o)
+		return nil
+	}
+}
+
+// ThatAreUp restricts the set of instances returned to only those with status UP.
+//
+// Combining the options produced by this function with those produced by WithStatus applies their
+// logical disjunction.
+func ThatAreUp() InstanceQueryOption {
+	return makeUncheckedStatusMatchingOption(UP)
+}
+
+// Shuffled requests randomizing the order of the sequence of instances returned, using the default
+// shared rand.Source.
+func Shuffled() InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		o.intn = rand.Intn
```

Comment:
I included (optional) shuffling to allow closer mimicking of how the Java `DiscoveryClient` class works. When one retrieves instances either from a VIP address or an application, it _automatically_ shuffles them. Well, to be more precise, each time that `DiscoveryClient` receives a set of instances from the server, it shuffles them and stores them in memory. `DiscoveryClient` answers subsequent queries against that same set of instances in the same order.

Here are a few code references to prove this assertion:
- [`DiscoveryClient#fetchRegistry()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java#L905) [calls `DiscoveryClient#getAndUpdateDelta()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java#L929)
- [`DiscoveryClient#getAndUpdateDelta()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java#L1043) [calls `DiscoveryClient#updateDelta()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java#L1061)
- [`DiscoveryClient#updateDelta()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java#L1167) [calls `Applications#shuffleInstances()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java#L1212)
- [`Applications#shuffleInstances()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/shared/Applications.java#L413) eventually [calls `Collections#shuffle()`](https://github.com/Netflix/eureka/blob/fed8381afae39038317b34b6fea359b90bd6f871/eureka-client/src/main/java/com/netflix/discovery/shared/Applications.java#L493)

Here, I didn't want to force callers to _disable_ automatic shuffling, but I wanted to provide the tools to closely match what `DiscoveryClient` does. It's germane to the problem domain—as the `DiscoveryClient` authors had decided—to want to ensure that different Eureka clients are not always choosing the same "first" instance from the set of those available. I figured that while it's just as possible to shuffle the order outside of _fargo_, doing so isn't as easy as a library call away.